cc/dual_net/dual_net.h

```diff
@@ -52,6 +56,118 @@ class DualNet {
   using StoneFeatures = std::array<float, kNumStoneFeatures>;
   using BoardFeatures = std::array<float, kNumBoardFeatures>;
 
+  struct Output {
+    std::array<float, kNumMoves> policy;
+    float value;
+  };
+
+  struct Result {
+    std::vector<Output> outputs;
+    std::string model;
+  };
+
+  class Service {
```

Comment:
Can you explain why the Service and Client classes are needed? This code structure has pushed a bunch of complexity down into the client classes, which I'd like to avoid. It seems to me that IncrementClientCount, DecrementClientCount and FlushClient are implementation details of batching and shouldn't be part of any public API.

I think you might have made life a little difficult for yourself by eliminating the factory classes: having the factory classes was useful because that was a place where one could put shared state.

When I was originally thinking about cleaning up the batching code, I imagined that there would be a BatchingFactory that wrapped a TfDualNetFactory (or whatever). The BatchingFactory::New would then simply return a client of the batcher that implemented the existing DualNet interface and the client refcounting and flushing would be implementation details of that DualNet subclass. That way none of the calling code would have to change at all: you'd just implement a batcher, batching factory and add the relevant code to factory.cc.

Reply:
The asynchronous playing of multiple games from a single thread requires some this complexity (in particular, that the client is shared between multiple players).

But I fully agree that I did not do a good job at hiding this complexity when it's not really needed. I hope the latest commit improved on it. 

The client count and flushing is still exposed in the service, but now transparently handled by the client. I think going all the way with an abstract client and an abstract factory isn't really warranted, but definitely possible.

Reply:
Playing multiple games from a single thread or multiple threads seems orthogonal to batching to me. In both cases, it seems like each player could continue to just have their own DualNet instance: a lightweight class that pushes requests to the batcher and waits for them to complete, much like how the previous InferenceServer worked. I don't understand why you need Client and Service in addition to DualNet.

Unfortunately, I'm very bad at trying to explain ideas that are in my head, so I resurrected my original attempt at extracting the batcher last night: https://github.com/tommadams/minigo/commit/291abff7061e88ab1039d9d080a4b42db834f9c7
It's just a sketch really but it seems to work and doesn't require the extra Client and Service classes.

I asked this before offline, but do you have benchmarks to demonstrate that playing multiple games from a single thread is significantly faster than just launching one thread per game? The setup you're targeting (a single process running on eight V100s, as I understand) is quite far removed from what we and hobbyist Minigo users run on, I'd like to understand whether the performance wins justify the non-trivial complexity added with async inference.

Even running your master branch code with the number of threads equal to the number of parallel games would be sufficient. I tried this on my desktop machine and didn't see any significant performance difference for 512 parallel games, but then I only have a fairly modest GPU.

Reply:
While out walking my dog, it occurred to me that this comment could be misinterpreted: I didn't mean to suggest that the batching code had to be structured or written a certain way, just that I don't believe that the Service and Client classes are strictly necessary.

This code will probably be read orders of magnitudes more times than it is written, so I'd like to keep it as simple as possible.

Reply:
Thanks Tom for the comments. I ran some experiments on my totally butchered repo. With a fake net, I got 38 sec on 16 threads vs 49 on 512 threads. So there is some improvement but it might not be relevant in the grand scheme. Let's put this aside for now, all code is back to synchronous execution in my latest commit.

Regarding the client count: it might be easier to understand why we need it if you consider the case of playing multiple eval games. Each game has two clients, one for each model. If each dual net waits for all clients to request an inference batch, they will deadlock each other.

With my latest commit. The service is now what the factory was in the original implementation. I didn't really see the value of the extra factory class when the service interface is very small. But you might want to convince me to rename the classes anyway ;-)
The client is an ABC with the client count being a private implementation detail of the batching client.